providers/gce/gce_loadbalancer_internal.go

```diff
@@ -700,7 +719,7 @@ func (g *Cloud) ensureInternalInstanceGroups(name string, nodes []*v1.Node) ([]s
 				igLinks = append(igLinks, ig.SelfLink)
 			}
 		} else {
-			igLink, err := g.ensureInternalInstanceGroup(name, zone, nodes)
+			igLink, err := g.ensureInternalInstanceGroup(name, zone, nodes, zonedNodes[""])
```

Comment:
How do we know that all Nodes with an empty zone belong to the zone being ensured here?

Specifically:

```
igLink, err := g.ensureInternalInstanceGroup(name, zone, nodes, zonedNodes[""])
                                                   ^^^^         ^^^^^^^^^^^^^^
                                                    A            B
```

That B all belong in zone A.

Reply:
so the logic here is that zonedNodes contain nodes that 'should not be removed from Instance Groups'.

What this means is that if, a node is attached to an IG, but for some reason it's zone assignment disappears, don't remove it from the group.
nodes in `emptyZoneNodes` will not be added, it's just to keep them from being removed.

inside `ensureInternalInstanceGroup` there is this part `removeNodes := gceNodes.Difference(kubeNodes).Difference(emptyZoneNodesNames).List()` - when determining which nodes to remove, skip the empty zones one.

In general we have many issues that start with nodes becoming 'temporarily' unready, or missing some data like zone info. Then LB controllers start removing them and we get outages.